templates/model/spec_test.go

```diff
@@ -0,0 +1,349 @@
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/abcxyz/pkg/testutil"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+func TestSpecUnmarshal(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name             string
+		in               string
+		want             *Spec
+		wantUnmarshalErr string
+		wantValidateErr  string
+	}{
+		{
+			name: "simple_template_should_succeed",
+			in: `apiVersion: 'abcxyz.dev/cli/v1alpha1'
+kind: 'Template'
+
+desc: 'A simple template that just prints and exits'
+inputs:
+- name: 'person_name'
+  required: false  # The default is required=true
+  desc: 'An optional name of a person to greet'
+
+steps:
+- desc: 'Print a message'
+  action: 'print'
+  params:
+    message: 'Hello, {{.or .person_name "World"}}'`,
+			want: &Spec{
+				APIVersion: String{Val: "abcxyz.dev/cli/v1alpha1"},
+				Kind:       String{Val: "Template"},
+
+				Desc: String{Val: "A simple template that just prints and exits"},
+				Inputs: []*Input{
+					{
+						Name:     String{Val: "person_name"},
+						Desc:     String{Val: "An optional name of a person to greet"},
+						Required: Bool{Val: false},
+					},
+				},
+				Steps: []*Step{
+					{
+						Desc:   String{Val: "Print a message"},
+						Action: String{Val: "print"},
+						Print: &Print{
+							Message: String{Val: `Hello, {{.or .person_name "World"}}`},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "validation_of_children_should_occur_and_fail",
+			in: `desc: 'A simple template that just prints and exits'
+inputs:
+- name: 'person_name'
+
+steps:
+- desc: 'Print a message'
+  action: 'print'
+  params:
+    message: 'Hello, {{.or .person_name "World"}}'`,
+			wantValidateErr: `invalid config near line 3 column 3: field "desc" is required`,
+		},
+		{
+			name: "check_required_fields",
+			in:   "inputs:",
+			wantValidateErr: `invalid config near line 1 column 1: field "apiVersion" is required
+invalid config near line 1 column 1: field "kind" is required
+invalid config near line 1 column 1: field "desc" is required
+invalid config near line 1 column 1: field "steps" is required`,
+		},
+
+		{
+			name: "unknown_field_should_fail",
+			in: `apiVersion: 'abcxyz.dev/cli/v1alpha1'
+kind: 'Template'
+
+desc: 'A simple template that just prints and exits'
+inputs:
+- name: 'person_name'
+  required: false  # The default is required=true
+  desc: 'An optional name of a person to greet'
+not_a_real_field: 'oops'
+
+steps:
+- desc: 'Print a message'
+  action: 'print'
+  params:
+    message: 'Hello'`,
+			wantUnmarshalErr: `invalid config near line 9 column 1: unknown field name "not_a_real_field"`,
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			got := &Spec{}
+			dec := NewDecoder(strings.NewReader(tc.in))
+			if err := dec.Decode(got); tc.wantUnmarshalErr != "" || err != nil {
+				if diff := testutil.DiffErrString(err, tc.wantUnmarshalErr); diff != "" {
```

Comment:
I might be misunderstanding the branching, but I think this misses a case where we want to assert where the marshall error is expected by the error is nil? I think the differrstring helper is smart enough to handle this if you just pass the values directly:

```go
err := dec.Decode(got)
if diff := testutil.DiffErrString(err, tc.wantUnmarshalErr); diff != nil {
  // ...
}
```

(throughout)

Reply:
But I want to skip the rest of the rest if there was an error, or if there was an expected error, or both. I don't want to try to Validate() if unmarshaling failed, regardless of whether hte failure was expected.

Reply:
What I tend to do is to return on err. I find it slightly cleaner

```go
err := dec.Decode(got)
if diff := testutil.DiffErrString(err, tc.wantUnmarshalErr); diff != "" {
  // ...
}
if err != nil {
  return
}
```

Reply:
That sounds good to me. Switched to Chen's suggestion.